### PR TITLE
Remove duplicate toUpperCase() calls

### PR DESCRIPTION
**Pull request by nyurik**

### CODE REVIEW

**Comment by SuperSuz** [comment]
I am not sure what is going on. I am learning to be a part of GitHub.
I am going to read the hello guide

**Comment by nyurik** [comment]
@shoeper I am not able to merge the changes in this repo. Either you can set up so that I can self-merge after approvals, or you need to merge it yourself. Thx!

P.S.  I think all `toUpperCase()` should be changed to `toUpperCase(Locale.ROOT)` because some tools may complain about locale ambiguity (even though in this case it shouldn't matter, ROOT makes it clear)

**Comment by nyurik** [comment]
@SuperSuz or @shoeper - are there any other people that need to review this code, or can you merge it? Thx!

**Comment by shoeper** [comment]
Sorry, I don't really know.

I cannot merge it.

**Comment by nyurik** [comment]
Looking at commit history, @drinckes and @JonMcPherson merged things recently.

File: java/com/google/openlocationcode/OpenLocationCode.java [hunk 1]
```diff
@@ -496,7 +498,7 @@ public static boolean isValidCode(String code) {
     if (code == null || code.length() < 2) {
       return false;
     }
-    code = code.toUpperCase();
+    code = code.toUpperCase(Locale.ROOT);
```
Comment (JonMcPherson) [inline java/com/google/openlocationcode/OpenLocationCode.java]: If you wanted to remove _all_ redundant calls of `toUpperCase()`, it could be modified to something like this
```
public static boolean isValidCode(String code) {
    return isValidCodeUpperCase(code.toUpperCase(Locale.ROOT));
}

private static boolean isValidCodeUpperCase(String code) {
    // do validity checks without calling code.toUpperCase() redundantly
}
```
And the constructor would instead call `!isValidCodeUpperCase(newCode)`.
Just a minor improvement...

**Comment by JonMcPherson** [comment]
Also, @drinckes is who you want for merging

**Comment by shoeper** [comment]
Maybe @zongweil can help.

File: java/com/google/openlocationcode/OpenLocationCode.java [hunk 2]
```diff
@@ -153,11 +154,12 @@ public double getEastLongitude() {
    * @constructor
   */
   public OpenLocationCode(String code) throws IllegalArgumentException {
-    if (!isValidCode(code.toUpperCase())) {
```
Comment (zongweil) [inline java/com/google/openlocationcode/OpenLocationCode.java]: We can get rid of the code.toUpperCase() here all together, since we're calling it inside of isValidCode().

File: java/com/google/openlocationcode/OpenLocationCode.java (hunk 1 above)
Comment (zongweil) [reply java/com/google/openlocationcode/OpenLocationCode.java]: Thanks for pointing out the redundant call. It's so redundant, in fact, that I think we can just remove the one in the constructor since the code will be cast to uppercase in isValidCode().

File: java/com/google/openlocationcode/OpenLocationCode.java [hunk 3]
```diff
@@ -153,11 +154,12 @@ public double getEastLongitude() {
...
+    String newCode = code.toUpperCase(Locale.ROOT);
+    if (!isValidCode(newCode)) {
       throw new IllegalArgumentException(
           "The provided code '" + code + "' is not a valid Open Location Code.");
     }
-    this.code = code.toUpperCase();
+    this.code = newCode;
```
Comment (zongweil) [inline java/com/google/openlocationcode/OpenLocationCode.java]: You can then leave this line as:

this.code = code.toUpperCase(Locale.ROOT);

File: java/com/google/openlocationcode/OpenLocationCode.java (hunk 2 above)
Comment (IDisposable) [reply java/com/google/openlocationcode/OpenLocationCode.java]: since this is public, can we make that assumption and document it the constructor comments?

**Comment by nyurik** [comment]
@zongweil I merged the code and applied your suggestion.

**Comment by nyurik** [comment]
I have done another pass at minor optimizations -- the code now has a `toValidCode` which returns normalized code or null if it fails. This way constructor can use it directly, without doing duplicate `toUpperCase`.

File: java/src/main/java/com/google/openlocationcode/OpenLocationCode.java [hunk 4]
```diff
@@ -555,79 +556,89 @@ public String toString() {
...
       // First longitude character can only have first 18 values.
       if (CODE_ALPHABET.indexOf(code.charAt(1)) > 17) {
-        return false;
+        return null;
       }
     }
 
     // Check the characters before the separator.
-    boolean paddingStarted = false;
+    boolean paddingStarted = null;
```
Comment (sonyaa) [inline java/src/main/java/com/google/openlocationcode/OpenLocationCode.java:601]:
```suggestion
    boolean paddingStarted = false;
```

This breaks the tests ("incompatible types: \<null\> cannot be converted to boolean").

File: java/src/main/java/com/google/openlocationcode/OpenLocationCode.java [hunk 5]
```diff
@@ -555,34 +556,44 @@ public String toString() {
...
+    return toValidCode(code) != null;
+  }
+
+  /**
+   * Normalizes Open Location, or returns `null` if the code is not valid.
+   *
+   * @param code The code to check and normalize.
+   * @return Normalized code if it is a valid full or short code, null otherwise.
+   */
+  private static String toValidCode(String code) {
```
Comment (phebehp) [inline java/src/main/java/com/google/openlocationcode/OpenLocationCode.java:568]: I think returning an Optional<String> instead of using null to signify invalid would be slightly cleaner here, what do you think?

File: java/com/google/openlocationcode/OpenLocationCode.java (hunk 2 above)
Comment (phebehp) [reply java/com/google/openlocationcode/OpenLocationCode.java]: I don't understand your question here, can you rephrase? What assumption are you thinking we need to document?

**Comment by drinckes** [comment]
Pinging everyone especially @nyurik to see if we can get the comments all resolved.
